File: day23.py
#!/usr/bin/env pipenv run python

import collections
import itertools
import re

from get_input import get_input, line_parser

class Point(collections.namedtuple("Point", ["x", "y", "z"])):
    def __repr__(self):
        return f"({self.x}, {self.y}, {self.z})"

    def __add__(self, other):
        if not isinstance(other, Point):
            return NotImplemented
        return type(self)(self.x + other.x, self.y + other.y, self.z + other.z)

    def distance(self, other):
        return abs(self.x - other.x) + abs(self.y - other.y) + abs(self.z - other.z)

class Bot(collections.namedtuple('Bot', ('radius', 'position'))):
    def __new__(cls, x, y, z, r):
        return super().__new__(cls, int(r), Point(int(x), int(y), int(z)))

def bots_in_range(position, radius, bots):
    for bot in bots:
        if position.distance(bot.position) <= radius:
            yield bot

def bots_in_box_range(position, radius, bots):
    total = 0
    for bot in bots:
        if position.distance(bot.position) <= radius + bot.radius:
            total += 1
    return total

def part1(bots):
    """Solution to part 1"""
    big_bot = max(bots, key=lambda b: b.radius)
    return sum(1 for b in bots_in_range(big_bot.position, big_bot.radius, bots))

def part2(bots):
    """Solution to part 2"""
    position = Point(0, 0, 0)
    max_radius = 1
    while bots_in_box_range(position, max_radius, bots) < len(bots):
        max_radius *= 2

    positions = set((position,))
    radius = max_radius
    while radius > 0:
        radius //= 2
        new_positions = set()
        seen = set()
        best_count = None
        for position in positions:
            for diff in itertools.chain(
                    ((0, 0, 0),),
                    itertools.product(*((0, radius, -radius), ) * 3)):
                new_position = Point(*diff) + position
                if new_position in seen:
                    continue
                seen.add(new_position)
                new_count = bots_in_box_range(new_position, radius, bots)
                if best_count is None or new_count > best_count:
                    new_positions.clear()
                    best_count = new_count
                if new_count == best_count:
                    new_positions.add(new_position)
        positions = new_positions
    return min(p.distance(Point(0, 0, 0)) for p in positions)

PART1_TEST = """pos=<0,0,0>, r=4
pos=<1,0,0>, r=1
pos=<4,0,0>, r=3
pos=<0,2,0>, r=1
pos=<0,5,0>, r=3
pos=<0,0,3>, r=1
pos=<1,1,1>, r=1
pos=<1,1,2>, r=1
pos=<1,3,1>, r=1"""


PART2_TEST = """pos=<10,12,12>, r=2
pos=<12,14,12>, r=2
pos=<16,12,12>, r=4
pos=<14,14,14>, r=6
pos=<50,50,50>, r=200
pos=<10,10,10>, r=5"""


def parse(line):
    return Bot(*re.match(r'pos=<(-?\d+),(-?\d+),(-?\d+)>, r=(\d+)', line).groups())

if __name__ == '__main__':
    assert part1(line_parser(PART1_TEST, parse=parse)) == 7
    bot_list = line_parser(get_input(day=23, year=2018), parse=parse)
    print("Part 1: {}".format(part1(bot_list)))
    assert part2(line_parser(PART2_TEST, parse=parse)) == 36
    print("Part 2: {}".format(part2(bot_list)))
    # 111960222

File: day23.rs
extern crate regex;

use regex::Regex;
use std::collections::HashSet;

use common::get_input;
use common::point::Point3d;

type Input = Vec<Bot>;

struct Bot {
    position: Point3d<i32>,
    radius: usize,
}

impl Bot {
    fn from_line(line: &str) -> Bot {
        let re = Regex::new(r"pos=<(-?\d+),(-?\d+),(-?\d+)>, r=(\d+)").unwrap();
        let cap = re.captures(&line).unwrap();
        Bot {
            position: Point3d::new(
              cap.get(1).unwrap().as_str().parse::<i32>().unwrap(),
              cap.get(2).unwrap().as_str().parse::<i32>().unwrap(),
              cap.get(3).unwrap().as_str().parse::<i32>().unwrap(),
            ),
            radius: cap.get(4).unwrap().as_str().parse::<usize>().unwrap(),
        }
    }
}


fn part1(input: &Input) -> usize {
    let big_bot = input.iter().max_by_key(|b| b.radius).unwrap();
    input.iter().filter(|b| 
        b.position.distance(&big_bot.position) < big_bot.radius
    ).count()
}

fn bot_in_box_range(bots: &Vec<Bot>, position: &Point3d<i32>, radius: usize) -> usize {
    bots.iter()
        .filter(|b| position.distance(&b.position) <= radius + b.radius)
        .count()
}



fn steps(radius: usize) -> Vec<Point3d<i32>> {
    let values = [0i32, radius as i32, -(radius as i32)];
    let mut indexs = [0; 3];
    let mut result: Vec<Point3d<i32>> = Vec::new();
    loop {
        result.push(Point3d::new(
            values[indexs[0]],
            values[indexs[1]],
            values[indexs[2]]
        ));
        for i in 0..3 {
            indexs[i] += 1;
            if indexs[i] < values.len() {
                break
            } else if i+1 < indexs.len() {
                indexs[i] = 0;
            } else {
                return result
            }
        }
    }
}

fn part2(input: &Input) -> usize {
    let position = Point3d::new(0, 0, 0);
    let mut max_radius = 1;
    while bot_in_box_range(&&input, &position, max_radius) < input.len() {
        max_radius *= 2;
    }
    let mut positions = HashSet::new();
    positions.insert(position);
    let mut radius = max_radius;
    while radius > 0 {
        radius /= 2;
        let mut new_positions = HashSet::new();
        let mut seen = HashSet::new();
        let mut best_count = None;
        for &position in positions.iter() {
            for diff in steps(radius) {
                let new_position = diff + position;
                if seen.contains(&new_position) {
                    continue;
                }
                seen.insert(new_position);
                let new_count = bot_in_box_range(input, &new_position, radius);
                if best_count.is_none() || best_count.unwrap() < new_count {
                    new_positions.clear();
                    best_count = Some(new_count);
                }
                if Some(new_count) == best_count {
                    new_positions.insert(new_position);
                }
            }
        }
        positions = new_positions;
    }
    positions.iter().map(|p| p.distance(&Point3d::new(0,0,0))).min().unwrap()
}

fn parse(input: String) -> Input {
    let mut bots = Vec::new();
    for line in input.trim().split('\n') {
        bots.push(Bot::from_line(line));
    }
    bots
}

fn main() {
    let input = parse(get_input(23, 2018));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
